distributed/tests/test_worker_state_machine.py

```diff
@@ -1351,3 +1351,201 @@ def test_transfer_incoming_metrics(ws):
+        GatherDepSuccessEvent(
+            worker=ws2, data={"a": 123}, total_nbytes=100, stimulus_id="s2"
+        )
+    )
+    assert instructions == [
+        AddKeysMsg.match(keys=["a"], stimulus_id="s2"),
+        GatherDep.match(worker=ws3, to_gather={"b"}, stimulus_id="s2"),
+    ]
+    assert ws.tasks["b"].state == "flight"
+
+
+def test_throttling_incoming_transfer_on_transfer_bytes_same_worker(ws):
+    ws.transfer_incoming_count_limit = 100
+    ws.transfer_incoming_bytes_limit = 250
+    ws.transfer_incoming_bytes_throttle_threshold = 1
+    ws2 = "127.0.0.1:2"
+    ws.handle_stimulus(
+        ComputeTaskEvent.dummy(
+            "d",
+            who_has={"a": [ws2], "b": [ws2], "c": [ws2]},
+            nbytes={"a": 100, "b": 100, "c": 100},
+            stimulus_id="s1",
+        )
+    )
+    assert ws.tasks["a"].state == "fetch"
+    assert ws.tasks["b"].state == "flight"
+    assert ws.tasks["c"].state == "flight"
+
+    # NOTE: We do not compare instructions since their sorting is random
+    ws.handle_stimulus(
+        GatherDepSuccessEvent(
+            worker=ws2, data={"b": 123, "c": 123}, total_nbytes=200, stimulus_id="s2"
+        )
+    )
+    assert ws.tasks["a"].state == "flight"
+    assert ws.tasks["b"].state == "memory"
+    assert ws.tasks["c"].state == "memory"
+
+
+def test_throttling_incoming_transfer_on_transfer_bytes_different_workers(ws):
+    ws.transfer_incoming_count_limit = 100
+    ws.transfer_incoming_bytes_limit = 150
+    ws.transfer_incoming_bytes_throttle_threshold = 1
+    ws2 = "127.0.0.1:2"
+    ws3 = "127.0.0.1:3"
+    ws.handle_stimulus(
+        ComputeTaskEvent.dummy(
+            "c",
+            who_has={"a": [ws2], "b": [ws3]},
+            nbytes={"a": 100, "b": 100},
+            stimulus_id="s1",
+        )
+    )
+    assert ws.tasks["a"].state == "flight"
+    assert ws.tasks["b"].state == "fetch"
+
+    instructions = ws.handle_stimulus(
+        GatherDepSuccessEvent(
+            worker=ws2, data={"a": 123}, total_nbytes=100, stimulus_id="s2"
+        )
+    )
+    assert instructions == [
+        AddKeysMsg.match(keys=["a"], stimulus_id="s2"),
+        GatherDep.match(worker=ws3, to_gather={"b"}, stimulus_id="s2"),
+    ]
+    assert ws.tasks["b"].state == "flight"
+
+
+def test_do_not_throttle_connections_while_below_threshold(ws):
+    ws.transfer_incoming_count_limit = 1
+    ws.transfer_incoming_bytes_limit = 200
+    ws.transfer_incoming_bytes_throttle_threshold = 50
+    ws2 = "127.0.0.1:2"
+    ws3 = "127.0.0.1:3"
+    ws4 = "127.0.0.1:4"
+    instructions = ws.handle_stimulus(
+        ComputeTaskEvent.dummy(
+            "b",
+            who_has={"a": [ws2]},
+            nbytes={"a": 1},
+            stimulus_id="s1",
+        )
+    )
+    assert instructions == [
+        GatherDep.match(
+            worker=ws2,
+            to_gather={"a"},
+            stimulus_id="s1",
+        ),
+    ]
+    assert ws.tasks["a"].state == "flight"
```

Comment:
I think you should either assert the state or the instruction. Both is redundant.

Reply:
Removed the instruction checks